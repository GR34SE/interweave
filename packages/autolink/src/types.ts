import React from 'react';

export interface LinkProps {
  children: NonNullable<React.ReactNode>;
  href?: string;
  newWindow?: boolean;
  onClick?: React.MouseEventHandler<HTMLAnchorElement>;
}

export interface EmailProps extends LinkProps {
  email: string;
}

export interface EmailMatch {
  email: string;
  emailParts: {
    host: string;
    username: string;
  };
}

export interface HashtagProps extends LinkProps {
  encodeHashtag?: boolean;
  hashtag: string;
  hashtagUrl?: string | ((hashtag: string) => string);
  preserveHash?: boolean;
}

export interface HashtagMatch {
  hashtag: string;
}

export interface UrlProps extends LinkProps {
  href: string;
}

export interface UrlMatch {
  url: string;
  urlParts: {
    auth: string;
    fragment: string;
    host: string;
    path: string;
    port: string | number;
    query: string;
    scheme: string;
  };
}

export interface UrlMatcherOptions {
  customTLDs?: string[];
  validateTLD?: boolean;
}
