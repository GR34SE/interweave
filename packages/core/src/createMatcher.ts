import React from 'react';

export type ElementFactory<Match, Props> = (
  content: NonNullable<React.ReactNode>,
  match: Match,
  props: Props,
) => React.ReactElement;

export interface MatchResult {
  index: number;
  length: number;
  match: string;
  matches: string[];
  valid: boolean;
  value: string;
  void: boolean;
}

export type MatchHandler<Match> = (value: string) => (MatchResult & { params: Match }) | null;

export interface Matcher<Match, Props> {
  factory: ElementFactory<Match, Props>;
  greedy: boolean;
  match: MatchHandler<Match>;
}

export type OnMatch<T> = (result: MatchResult) => T | null;

export interface MatcherOptions<Match> {
  greedy?: boolean;
  tagName: string;
  void?: boolean;
  // onAfterParse?: (content: Node[], props: Props) => Node[];
  // onBeforeParse?: (content: string, props: Props) => string;
  onMatch: OnMatch<Match>;
}

export default function createMatcher<Match, Props>(
  pattern: string | RegExp,
  options: MatcherOptions<Match>,
  factory: ElementFactory<Match, Props>,
): Matcher<Match, Props> {
  return {
    factory,
    greedy: options.greedy ?? false,
    match(value) {
      const matches = value.match(pattern instanceof RegExp ? pattern : new RegExp(pattern, 'i'));

      if (!matches) {
        return null;
      }

      const result: MatchResult = {
        index: matches.index!,
        length: matches[0].length,
        match: matches[0],
        matches,
        valid: true,
        value,
        void: options.void ?? false,
      };
      const params = options.onMatch(result);

      // Allow callback to intercept the result
      if (params === null) {
        return null;
      }

      return {
        params,
        ...result,
      };
    },
  };
}
