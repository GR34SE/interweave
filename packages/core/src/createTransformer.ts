import { WildTagName, InferElement, Transformer, TransformerFactory } from './types';

export default function createTransformer<
  K extends WildTagName,
  Element = InferElement<K>,
  Props = {}
>(tagName: K, factory: TransformerFactory<Element, Props>): Transformer<Element, Props> {
  return {
    factory,
    tagName,
  };
}
