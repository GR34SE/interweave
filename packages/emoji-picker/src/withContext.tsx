import React from 'react';
import { Context as EmojiContext } from './types';

export const Context = React.createContext({
  classNames: {},
  messages: {},
});

export interface WithContextProps {
  context: EmojiContext;
}

export default function withContext<Props extends object>(
  Component: React.ComponentType<Props & WithContextProps>,
): React.ComponentType<Props> {
  function WithContextWrapper(props: Props) {
    return (
      <Context.Consumer>
        {/* istanbul ignore next */ context => <Component {...props} context={context} />}
      </Context.Consumer>
    );
  }

  WithContextWrapper.displayName = `withEmojiContext(${Component.displayName || Component.name})`;

  return WithContextWrapper;
}
