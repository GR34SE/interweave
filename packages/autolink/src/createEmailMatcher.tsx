import React from 'react';
import createMatcher, { ElementFactory, MatchResult } from 'interweave/src/createMatcher';
import Email from './Email';
import { EMAIL_PATTERN } from './constants';
import { EmailMatch } from './types';

function onMatch({ matches }: MatchResult): EmailMatch {
  return {
    email: matches[0],
    emailParts: {
      host: matches[2],
      username: matches[1],
    },
  };
}

export default function createEmailMatcher<Props>(factory?: ElementFactory<EmailMatch, Props>) {
  return createMatcher<EmailMatch, Props>(
    EMAIL_PATTERN,
    { onMatch, tagName: 'a' },
    factory || ((content, { email }) => <Email email={email}>{content}</Email>),
  );
}
