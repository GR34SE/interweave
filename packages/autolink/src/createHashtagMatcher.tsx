import React from 'react';
import createMatcher, { ElementFactory, MatchResult } from 'interweave/src/createMatcher';
import Hashtag from './Hashtag';
import { HASHTAG_PATTERN } from './constants';
import { HashtagMatch } from './types';

function onMatch({ matches }: MatchResult): HashtagMatch {
  return {
    hashtag: matches[0],
  };
}

export default function createHashtagMatcher<Props>(factory?: ElementFactory<HashtagMatch, Props>) {
  return createMatcher<HashtagMatch, Props>(
    HASHTAG_PATTERN,
    { onMatch, tagName: 'a' },
    factory || ((content, { hashtag }) => <Hashtag hashtag={hashtag}>{content}</Hashtag>),
  );
}
