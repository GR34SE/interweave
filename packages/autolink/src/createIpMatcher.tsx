import React from 'react';
import createMatcher, { ElementFactory } from 'interweave/src/createMatcher';
import Url from './Url';
import { onMatch } from './createUrlMatcher';
import { IP_PATTERN } from './constants';
import { UrlMatch } from './types';

export default function createIpMatcher<Props>(factory?: ElementFactory<UrlMatch, Props>) {
  return createMatcher<UrlMatch, Props>(
    IP_PATTERN,
    { onMatch, tagName: 'a' },
    factory || ((content, { url }) => <Url href={url}>{content}</Url>),
  );
}
