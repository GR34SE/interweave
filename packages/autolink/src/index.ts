/**
 * @copyright   2016-2019, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import createEmailMatcher from './createEmailMatcher';
import createHashtagMatcher from './createHashtagMatcher';
import createIpMatcher from './createIpMatcher';
import createUrlMatcher from './createUrlMatcher';
import Email from './Email';
import Hashtag from './Hashtag';
import Link from './Link';
import Url from './Url';

export {
  createEmailMatcher,
  createHashtagMatcher,
  createIpMatcher,
  createUrlMatcher,
  Email,
  Hashtag,
  Link,
  Url,
};

export * from './constants';
export * from './types';
