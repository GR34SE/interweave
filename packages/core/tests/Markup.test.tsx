import React from 'react';
import { render } from 'rut';
import Markup from '../src/Markup';
import Element from '../src/Element';
import { MOCK_MARKUP } from '../src/testUtils';

const options = { log: false, reactElements: false };

describe('Markup', () => {
  it('can change `tagName`', () => {
    const { root } = render(<Markup tagName="p" content="Foo Bar" />);

    expect(root.findOne(Element)).toHaveProp('tagName', 'p');
  });

  it('can use a fragment', () => {
    const { root } = render(<Markup tagName="fragment" content="Foo Bar" />);

    expect(root).toContainNode('Foo Bar');
  });

  it('allows empty `content` to be passed', () => {
    const { root } = render(<Markup content={null} />);

    expect(root.debug(options)).toMatchSnapshot();
  });

  it('will render the `emptyContent` if no content exists', () => {
    const empty = <div>Foo</div>;
    const { root } = render(<Markup content="" emptyContent={empty} />);

    expect(root).toContainNode(empty);
  });

  it('parses the entire document starting from the body', () => {
    const { root } = render(<Markup content={MOCK_MARKUP} />);

    expect(root.debug(options)).toMatchSnapshot();
  });

  it('converts line breaks', () => {
    const { root } = render(<Markup content={'Foo\nBar'} />);

    expect(root.debug(options)).toMatchSnapshot();
  });

  it('doesnt convert line breaks', () => {
    const { root } = render(<Markup content={'Foo\nBar'} disableLineBreaks />);

    expect(root.debug(options)).toMatchSnapshot();
  });

  it('doesnt convert line breaks if it contains HTML', () => {
    const { root } = render(<Markup content={'Foo\n<br/>Bar'} />);

    expect(root.debug(options)).toMatchSnapshot();
  });
});
