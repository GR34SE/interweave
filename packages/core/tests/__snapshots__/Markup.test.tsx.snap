// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Markup allows empty \`content\` to be passed 1`] = `"<div />"`;

exports[`Markup converts line breaks 1`] = `
"<div>
  Foo
  <br />
  Bar
</div>"
`;

exports[`Markup doesnt convert line breaks 1`] = `
"<div>Foo
Bar</div>"
`;

exports[`Markup doesnt convert line breaks if it contains HTML 1`] = `
"<div>
  Foo

  <br />
  Bar
</div>"
`;

exports[`Markup parses the entire document starting from the body 1`] = `
"<div>
  <main role=\\"main\\">
    
  Main content
  
    <div>
      
    
      <a href=\\"#\\">
        Link
      </a>
      
    
      <span className=\\"foo\\">
        String
      </span>
      
  
    </div>
    

  </main>
  

  <aside id=\\"sidebar\\">
    
  Sidebar content

  </aside>
</div>"
`;
